In [ ]:
from IPython.display import display, clear_output, HTML
import time

def countdown_timer(minutes):
    total_seconds = minutes * 60
    for seconds in range(total_seconds, 0, -1):
        mins, secs = divmod(seconds, 60)
        time_str = f"{mins:02}'{secs:02}''"
        clear_output(wait=True)
        # HTML with styling
        display(HTML(f'<div style="font-size: 24px; color: blue; font-weight: bold;">Time remaining: {time_str}</div>'))
        time.sleep(1)
    clear_output(wait=True)
    # Final message with different styling
    display(HTML('<div style="font-size: 24px; color: green; font-weight: bold;">Time\'s up!</div>'))

<div style="text-align: center;">
    <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/f8/Python_logo_and_wordmark.svg/972px-Python_logo_and_wordmark.svg.png?20210516005643" alt="The Python logo" style="width: 60%; max-width: 500px;">
</div>

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise

Insert a `Markdown` cell containing a heading: "Data Structures".
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung

Fügen Sie eine `Markdown`-Zelle mit einer Überschrift ein: "Datenstrukturen".
</div>
</div>

In [ ]:
#countdown
countdown_timer(1)

<div class="alert alert-block alert-info">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

We’ll explore many objects and their methods today. 

Don’t worry—it might seem like a lot of information, but many features are recurring, and by encountering them repeatedly, you'll start to recognize them over time. 

The goal is for you to build a mental map of what’s possible.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Wir werden heute viele Objekte und ihre Methoden untersuchen. 

Keine Sorge - es mag Ihnen wie eine Menge Informationen vorkommen, aber viele Funktionen tauchen immer wieder auf, und wenn Sie ihnen wiederholt begegnen, werden Sie sie mit der Zeit erkennen. 

Das Ziel ist, dass Sie sich ein Bild davon machen, was möglich ist.

</div>
</div>

<div class="alert alert-block alert-warning">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

# Lists 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

# Listen
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

A string is a *sequence*: a collection of objects (in that case, characters).
Lists are also an example of sequences.
Lists in Python are used to store ordered collections of values. You can create a list by placing elements inside square brackets, separated by commas:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Ein String ist eine *Sequenz*: eine Sammlung von Objekten (in diesem Fall von Zeichen).
Auch Listen sind ein Beispiel für Sequenzen.
Listen in Python werden verwendet, um geordnete Sammlungen von Werten zu speichern. Sie können eine Liste erstellen, indem Sie Elemente in eckige Klammern setzen, die durch Kommas getrennt sind:

</div>
</div>

In [16]:
# a list of numbers

prime_numbers = [2, 3, 5, 7, 11, 13]

type(prime_numbers)

list

In [ ]:
# a list of strings

planets = ['Mercury', 'Venus', 'Earth', 'Mars', 'Jupiter', 'Saturn', 'Uranus', 'Neptune']

In [ ]:
# a list of lists

hands = [
['J', 'Q', 'K'],
['2', '2', '2'],
['6', 'A', 'K']
]

# (it can also be written on one line, but it can get hard to read)
# hands = [['J', 'Q', 'K'], ['2', '2', '2'], ['6', 'A', 'K']]

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

The individual entries do not need to have the same type:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Die einzelnen Einträge müssen nicht den gleichen Typ haben:

</div>
</div>

In [ ]:
# a list of mixed data types

a = [3.2, 'hello', [4, 8, 17.1]] 

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise
Channel your inner Quentin Tarantino and create a Python list that captures the five key names from "Kill Bill's" notorious death list.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung
Lassen Sie Ihren inneren Quentin Tarantino sprechen und erstellen Sie eine Python-Liste, die die fünf wichtigsten Namen aus der berüchtigten Todesliste von "Kill Bill" enthält.

</div>
</div>

<div style="text-align: center;">
<img src="./figures/death_list_five.png" width="300">
</div>

In [ ]:
#countdown
countdown_timer(5)

In [ ]:
death_list_five = ['O-Ren Ishii', 'Vernita Green', 'Budd', 'Elle Driver', 'Bill']

death_list_five

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
Lists are *ordered*, which means that its elements can be referred to with unique integer indexes.
Therefore Lists can be indexed, sliced and iterated over just like strings.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
Listen sind *geordnet*, was bedeutet, dass auf ihre Elemente mit eindeutigen ganzzahligen Indizes verwiesen werden kann.
Daher können Listen genau wie Zeichenketten indiziert, zerlegt und iteriert werden.
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
## Indexing 

You can access individual elements `i` in a list using square brackets with the syntax `list[i]`. To retrieve the first element of a list, you would use index `0`, as Python uses zero-based indexing:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
## Indexierung 

 Sie können auf einzelne Elemente `i` in einer Liste zugreifen, indem Sie eckige Klammern mit der Syntax `list[i]` verwenden. Um das erste Element einer Liste abzurufen, verwenden Sie den Index `0`, da Python eine nullbasierte Indizierung verwendet:

</div>
</div>

<div style="text-align: center;">
<img src="./figures/ListIndex.png" width="400"> 
</div>

In [ ]:
planets = ['Mercury', 'Venus', 'Earth', 'Mars', 'Jupiter', 'Saturn', 'Uranus', 'Neptune']

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
#### Exercise
What is the first planet (closest to the Sun)?

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
#### Übung
Was ist der erste Planet (am nächsten zur Sonne)?

</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

When an entry in a list is itself a list, we can use a second index to access the elements of this: `list[i][j]`.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Wenn ein Eintrag in einer Liste selbst eine Liste ist, können wir einen zweiten Index verwenden, um auf die Elemente dieser Liste zuzugreifen: `list[i][j]`.

</div>
</div>

In [12]:
hands = [
['J', 'Q', 'K'],
['2', '2', '2'],
['6', 'A', 'K']
]

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
#### Exercise
Try to get the position of the element "A".

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
#### Übung
Versuchen Sie, die Position des Elements "A" zu ermitteln.

</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

In [14]:
hands[2][1]

'A'

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

Elements at the end of a list can be accessed using negative indices, starting with `-1` for the last element:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Auf Elemente am Ende einer Liste kann mit negativen Indizes zugegriffen werden, beginnend mit `-1` für das letzte Element:

</div>
</div>

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
#### Exercise
What is the last planet (farthest from the Sun)?

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
#### Übung
Welcher ist der letzte Planet (am weitesten von der Sonne entfernt)?

</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

In [ ]:
planets[-1]

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

The advantage of using negative list indices is that we don't need to know the length of the list to retrieve the last value, as its index will always be `-1` regardless of the list's length.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Der Vorteil der Verwendung negativer Listenindizes besteht darin, dass wir die Länge der Liste nicht kennen müssen, um den letzten Wert abzurufen, da sein Index unabhängig von der Länge der Liste immer `-1` ist.

</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
 
## Slicing 

Slicing allows you to access a subset of a list by specifying a start index, an end index, and an optional step. The syntax for slicing is:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Slicing 

Mit Slicing können Sie auf eine Teilmenge einer Liste zugreifen, indem Sie einen Startindex, einen Endindex und einen optionalen Schritt angeben. Die Syntax für Slicing lautet:

</div>
</div>

```python
list[start:end:step]        # returns items from start to (end-1) at intervals defined by step
```

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

Beware: the right slice index specifies where we want to stop slicing, and <b>is not</b> included in the returned slice.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Achtung: Der rechte Slice-Index gibt an, wo wir mit dem Slicen aufhören wollen, und <b>ist nicht</b> im zurückgegebenen Slice enthalten.

</div>
</div>

In [18]:
german_national_team = [
# Goalkeepers/Torhüter
"Manuel Neuer", 
"Marc-André ter Stegen", 
"Oliver Baumann",

# Defenders/Verteidiger
"Antonio Rüdiger", 
"Nico Schlotterbeck", 
"David Raum", 
"Waldemar Anton", 
"Jonathan Tah", 
"Robin Koch", 
"Maximilian Mittelstädt", 
"Benjamin Henrichs", 
"Joshua Kimmich",

# Midfielders/Mittelfeldspieler
"Ilkay Gündogan", 
"Jamal Musiala", 
"Leroy Sané", 
"Florian Wirtz", 
"Pascal Groß", 
"Emre Can", 
"Robert Andrich", 
"Chris Führich", 
"Toni Kroos",

# Forwards/Stürmer
"Kai Havertz", 
"Niclas Füllkrug", 
"Thomas Müller", 
"Maximilian Beier", 
"Deniz Undav"
]

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
#### Exercise

 Using list slicing, extract the goalkeepers, defenders, midfielders, and forwards from the `german_national_team` list. 
 
 Assign the results to variables with descriptive names for each group of players.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
#### Übung

Extrahieren Sie mithilfe von List Slicing die Torhüter, Verteidiger, Mittelfeldspieler und Stürmer aus der Liste `german_national_team`. 

Weisen Sie die Ergebnisse den Variablen mit beschreibenden Namen für jede Gruppe von Spielern zu.
</div>
</div>

In [ ]:
#countdown
countdown_timer(10)

In [19]:
goalkeepers = german_national_team[:3] 
print(goalkeepers)  

defenders = german_national_team[3:12] 
print(defenders)  

midfielders = german_national_team[12:21] 
print(midfielders)  

forwards = german_national_team[21:] 
print(forwards)  

['Manuel Neuer', 'Marc-André ter Stegen', 'Oliver Baumann']
['Antonio Rüdiger', 'Nico Schlotterbeck', 'David Raum', 'Waldemar Anton', 'Jonathan Tah', 'Robin Koch', 'Maximilian Mittelstädt', 'Benjamin Henrichs', 'Joshua Kimmich']
['Ilkay Gündogan', 'Jamal Musiala', 'Leroy Sané', 'Florian Wirtz', 'Pascal Groß', 'Emre Can', 'Robert Andrich', 'Chris Führich', 'Toni Kroos']
['Kai Havertz', 'Niclas Füllkrug', 'Thomas Müller', 'Maximilian Beier', 'Deniz Undav']


<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

The starting and ending indices in slicing are both optional.

-  If you omit the start index, it defaults to `0`, meaning the slice begins from the start of the list. 

-  If you omit the end index, it defaults to the length of the list, meaning the slice extends to the end of the list.

-  If you omit both the starting and ending indices but keep a value `n` for the step, you get every n-th value from the entire list.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Die Start- und Endindizes beim Slicing sind beide optional.

- Wenn Sie den Startindex weglassen, ist er standardmäßig `0`, was bedeutet, dass das Slicing am Anfang der Liste beginnt.

-  Wenn Sie den Endindex weglassen, wird er auf die Länge der Liste gesetzt, was bedeutet, dass das Slicing bis zum Ende der Liste reicht.

-  Wenn Sie sowohl den Start- als auch den Endindex weglassen, aber einen Wert `n` für den Schritt beibehalten, erhalten Sie jeden n-ten Wert der gesamten Liste.

</div>
</div>

In [ ]:
# Get every other planet from the beginning to the end
every_other = planets[::2]
print(every_other)  

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise

-  Create a list of the first 10 integers

-  Extract all odd numbers

-  Extract all even numbers

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung

-  Erstellen einer Liste mit den ersten 10 ganzen Zahlen

-  Extrahiere alle ungeraden Zahlen

-  Extrahiere alle geraden Zahlen

</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

In [ ]:
numbers = [1,2,3,4,5,6,7,8,9,10]

odd_numbers = numbers[::2]
print(odd_numbers)

even_numbers = numbers[1::2]
print(even_numbers)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Manipulating lists 

Lists are *mutable* (unlike strings!), which means you can change their contents directly. 

You can modify a list by assigning new values to specific indices or slices. 

For example, you can update individual elements or replace a range of elements using indexing or slicing.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Listen manipulieren

Listen sind *veränderlich* (im Gegensatz zu Strings!), d.h. ihr Inhalt kann direkt geändert werden. 

Sie können eine Liste ändern, indem Sie bestimmten Indizes oder Slices neue Werte zuweisen. 

So können Sie beispielsweise einzelne Elemente aktualisieren oder einen Bereich von Elementen durch Indizierung oder Slicing ersetzen.

</div>
</div>

In [ ]:
planets

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
### Replacing elements    
 To change the value of an individual element, we assign a new value to its index, similar to how we would assign a new value to a variable:
 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
###  Elemente ersetzen 
Um den Wert eines einzelnen Elements zu ändern, weisen wir seinem Index einen neuen Wert zu, ähnlich wie wir einer Variablen einen neuen Wert zuweisen würden:

</div>
</div>

In [ ]:
planets[3] = 'Tatooine'

planets

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
 You can use list slicing to replace several elements simultaneously:
 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
 Sie können die Listenaufteilung verwenden, um mehrere Elemente gleichzeitig zu ersetzen:

</div>
</div>

In [ ]:
planets[:2] = ['Krypton', 'Solaris']

planets

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
#### Exercise

Change the planet names back to those of our solar system using indexing and slicing.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
#### Übung

Ändern Sie die Planetennamen mit Hilfe von Indizierung und Slicing in die Namen unseres Sonnensystems zurück.

</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

If the element is a number, we can manipulate it mathematically:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Handelt es sich bei dem Element um eine Zahl, können wir es mathematisch manipulieren:

</div>
</div>

In [ ]:
some_numbers=[2, 2, 2, 2]

In [ ]:
some_numbers[0]=some_numbers[0]+7

some_numbers[1]=some_numbers[1]/3

some_numbers[2]=some_numbers[2]*2

some_numbers[3]=some_numbers[3]**12

some_numbers

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

### Adding elements    
 There are several methods for adding single elements to a list, depending on the desired position of the new element.
 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

### Elemente hinzufügen 
Es gibt mehrere Methoden, um einzelne Elemente an eine Liste anzufügen, je nachdem, an welcher Stelle das neue Element stehen soll.

</div>
</div>

#### `append()`

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

To add an element to the end of a list, we use the `append()` method:
 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Um ein Element am Ende einer Liste hinzuzufügen, verwenden wir die Methode `append()`:

</div>
</div>

In [ ]:
letters=["b", "c", "d", "f"]

In [ ]:
letters.append("g")
letters.append("h")

letters

#### `insert()`

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

 We use the `insert()` method to add new entries to the beginning or middle of a list. 
 
 This method is slightly more complex than `append()` because it requires two arguments rather than one. 
 
 The first argument specifies the position—using the list index—where the new element will be inserted, and the second argument is the new element itself:
 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Wir verwenden die Methode `insert()`, um neue Einträge am Anfang oder in der Mitte einer Liste hinzuzufügen. 

Diese Methode ist etwas komplizierter als `append()`, da sie zwei Argumente statt einem benötigt. 

Das erste Argument gibt die Position an, an der das neue Element eingefügt werden soll, und das zweite Argument ist das neue Element selbst:

</div>
</div>

In [ ]:
letters

In [ ]:
letters.insert(0, "a")

letters.insert(4, "e")

letters

#### `list_a + list_b`

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
 If we "add" two lists together, we create a new list containing the elements from each:
 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
 Wenn wir zwei Listen "addieren", erstellen wir eine neue Liste, die die Elemente der beiden Listen enthält:

</div>
</div>

In [ ]:
numbers = [1,2,3,4,5,6,7,8,9,10]

odd_numbers = numbers[::2]
print(odd_numbers)

even_numbers = numbers[1::2]
print(even_numbers)

In [ ]:
odd_numbers + even_numbers

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
 The elements are added in the order they appear; we will see in a moment how to sort them in a meaningful way.
 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
 Die Elemente werden in der Reihenfolge hinzugefügt, in der sie erscheinen; wir werden gleich sehen, wie man sie sinnvoll sortiert.

</div>
</div>

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
#### Exercise

The list `powers` in the code box below is intended to contain the first eight values of $2^n$ (i.e., for values of n ranging from 0 to 7 inclusive). However, the list currently has some missing elements and contains incorrect values:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
#### Übung

Die Liste der Potenzen `powers` in der unten stehenden Codebox soll die ersten acht Werte von $2^n$ enthalten (d. h. für Werte von n von 0 bis einschließlich 7). Allerdings fehlen in der Liste derzeit einige Elemente und sie enthält falsche Werte:
</div>
</div>

In [ ]:
powers=[1, 2, 5, 8, 33, 63]

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

- Change the value of any existing elements that are incorrect.

- Add any missing elements to ensure that the list contains the correct values.

</div>
<div style="width: 48%; line-height: 1.3; color: grey;">

- Ändern Sie den Wert aller vorhandenen Elemente, die nicht korrekt sind.

- Fügen Sie alle fehlenden Elemente hinzu, um sicherzustellen, dass die Liste die richtigen Werte enthält.

</div>
</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

In [ ]:
powers[2]=4
powers[4]=32
powers[5]=64
powers.append(128)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

### Removing elements    

 To remove elements from a list, we can use the `pop()` method by specifying the index of the element we want to remove:
 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

###  Elemente entfernen 

Um Elemente aus einer Liste zu entfernen, kann man die Methode `pop()` verwenden, indem man den Index des zu entfernenden Elements angibt:

</div>
</div>

In [ ]:
#Create a list called "vegetables" containing incorrect entries (fruits)
vegetables=["carrot", "orange", "broccoli", "onion", "strawberry", "asparagus", "apple"]

In [ ]:
#orange (list index 1) is a fruit - remove it!
vegetables.pop(1)

vegetables

In [ ]:
#strawberry is also a fruit - remove it!
vegetables.pop(3) #note that because we removed orange from the list, the index for strawberry has decreased by one

#apple is certainly not a vegetable - remove it!
vegetables.pop(-1) #we can also use negative indices to specify which element to remove

#print the correct list
vegetables

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Some useful functions 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Einige nützliche Funktionen
</div>
</div>

#### `len()`

In [21]:
planets = ["Mercury", "Venus", "Earth", "Mars", "Jupiter", "Saturn", "Uranus", "Neptune"]

len(planets)

8

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise
 How many players are in the german football team?

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung
 Wie viele Spieler sind in der deutschen Fußballmannschaft?
</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

#### `sorted()`

In [ ]:
sorted(planets)

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise

 Sort the outcome of `odd_numbers + even_numbers`.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung

 Sortieren Sie das Ergebnis von `ungerade_Zahlen + gerade_Zahlen`.
</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

#### `in` (membership test)

In [22]:
# Is Pluto a planet?
"Pluto" in planets

False

#### `sum()`

In [ ]:
prime_numbers = [2, 3, 5, 7, 11, 13]

sum(prime_numbers)

#### `min()`

In [ ]:
min(prime_numbers)

#### `max()`

In [ ]:
max(prime_numbers)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

**Lists** are highly versatile and widely used in Python for several reasons:

  - **Ordered collections:** Lists maintain the order of elements, allowing for indexing, slicing, and iteration. This makes them ideal for sequences where the order matters.

  - **Mutable:** Lists can be modified after creation, which means you can add, remove, or change elements as needed without creating a new list.
  
  - **Allow duplicates:** Lists can store duplicate elements, which is useful when the same value might need to appear multiple times.
  
  - **Heterogeneous data:** Lists can hold elements of different data types, such as integers, strings, or even other lists, making them flexible for various use cases.
  
  - **Dynamic resizing:** Lists can grow and shrink dynamically, meaning you don't have to specify the size of the list ahead of time.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

**Listen** sind sehr vielseitig und werden aus mehreren Gründen häufig verwendet:

  - **Geordnete Sammlungen:** Listen behalten die Reihenfolge der Elemente bei und ermöglichen Indizierung, Aufteilung und Iteration. Das macht sie ideal für Sequenzen, bei denen die Reihenfolge wichtig ist.
  
  - **Veränderbar:** Listen können nach der Erstellung geändert werden, d.h. Sie können Elemente hinzufügen, entfernen oder ändern, ohne eine neue Liste zu erstellen.
  
  - **Zulassen von Duplikaten:** Listen können doppelte Elemente speichern, was nützlich ist, wenn derselbe Wert mehrmals vorkommen muss.
  
  - **Heterogene Daten:** Listen können Elemente verschiedener Datentypen enthalten, wie z. B. Ganzzahlen, Zeichenketten oder sogar andere Listen, was sie für verschiedene Anwendungsfälle flexibel macht.
  
  - **Dynamische Größenanpassung:** Listen können dynamisch wachsen und schrumpfen, d.h. Sie müssen die Größe der Liste nicht im Voraus festlegen.

</div>
</div>

<div class="alert alert-block alert-warning">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

# Tuples 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

# Tupel
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

A tuple is an *immutable, ordered* sequence of objects (lists are mutable and ordered).


Tuples are defined by providing the sequence of elements between parentheses:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Ein Tupel ist eine *unveränderliche, geordnete* Folge von Objekten (Listen sind veränderbar und geordnet).


Tupel werden definiert, indem die Reihenfolge der Elemente zwischen Klammern angegeben wird:

</div>
</div>

In [ ]:
tuple = (0, 'a', 2.3)

tuple

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Indexing, Slicing and manipulation 

As an immutable data structure, assignment to a tuple is not allowed and appending and removing elements is not supported.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Indexierung, Slicing und Manipulation 

Da es sich um eine unveränderliche Datenstruktur handelt, ist eine Zuweisung an ein Tupel nicht zulässig, und das Anhängen und Entfernen von Elementen wird nicht unterstützt.

</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

**Tuples** are useful in Python for several reasons:


  - **Immutability:** Tuples are immutable, meaning their elements cannot be changed after creation. This makes them suitable for representing fixed collections of data that should not be modified, such as coordinates, days of the week, or settings. Immutability also ensures data integrity and can help prevent accidental changes.

  - **Performance:** Since tuples are immutable, they are more memory-efficient and can be faster to access compared to lists. If you need to store a collection of values that won’t change, tuples are a more optimized choice.

  - **Dictionary Keys:** Tuples can be used as keys in dictionaries because they are immutable. Lists, on the other hand, cannot be used as dictionary keys.

  - **Return Multiple Values:** Functions can return multiple values as tuples. This is a common pattern in Python when you want to return several results from a function at once.

  - **Data Structure:** Tuples often represent structured data, like a record in a database, where each element has a specific meaning (e.g., `(name, age, city)`).


In summary, tuples are important when you need an ordered, fixed collection of elements with benefits in performance, memory efficiency, and data integrity.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

**Tupel** sind in Python aus mehreren Gründen nützlich:


  - **Veränderbarkeit:** Tupel sind unveränderlich, d.h. ihre Elemente können nach der Erstellung nicht geändert werden. Dadurch eignen sie sich zur Darstellung fester Datensammlungen, die nicht geändert werden sollten, wie Koordinaten, Wochentage oder Einstellungen. Die Unveränderlichkeit stellt auch die Datenintegrität sicher und kann dazu beitragen, versehentliche Änderungen zu verhindern.

  - **Leistung:** Da Tupel unveränderlich sind, sind sie speichereffizienter und der Zugriff kann im Vergleich zu Listen schneller erfolgen. Wenn Sie eine Sammlung von Werten speichern müssen, die sich nicht ändern, sind Tupel die bessere Wahl.

  - **Wörterbuchschlüssel:** Tupel können als Schlüssel in Dictionaries (Wörterbüchern) verwendet werden, da sie unveränderlich sind. Listen hingegen können nicht als Wörterbuchschlüssel verwendet werden.

  - **Mehrere Werte zurückgeben:** Funktionen können mehrere Werte als Tupel zurückgeben. Dies ist ein gängiges Muster in Python, wenn man mehrere Ergebnisse von einer Funktion auf einmal zurückgeben möchte.

  - **Datenstruktur:** Tupel repräsentieren oft strukturierte Daten, wie einen Datensatz in einer Datenbank, wo jedes Element eine bestimmte Bedeutung hat (z.B. `(Name, Alter, Stadt)`).

Zusammenfassend lässt sich sagen, dass Tupel wichtig sind, wenn man eine geordnete, feste Sammlung von Elementen benötigt, die Vorteile in Bezug auf Leistung, Speichereffizienz und Datenintegrität bietet.
</div>
</div>

<div class="alert alert-block alert-warning">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

# Sets 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

# Sets
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

A Python set is a *mutable, unordered* collection of objects with no duplicate elements.


A non-empty set can be defined by providing the elements between braces (but `set = {}` defines a dictionary (see below), not the empty set!):

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Ein Python-Set ist eine *veränderliche, ungeordnete* Sammlung von Objekten, die keine doppelten Elemente enthält.


Ein nicht leeres Set kann definiert werden, indem die Elemente zwischen geschweiften Klammern angegeben werden (aber `set = {}` definiert ein Wörterbuch (siehe unten), nicht die leere Menge!):

</div>
</div>

In [15]:
set = {1, 1, 'a', 'b', 1, 2, 'a', 1.0}

set

{1, 2, 'a', 'b'}

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

What happened?
Sets automatically remove duplicate values, and `1.0` is treated as the same value as `1`.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Was ist passiert?
Sets entfernen automatisch doppelte Werte, und `1.0` wird als der gleiche Wert wie `1` behandelt.

</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Indexing and Slicing 

Sets cannot be indexed, since they are not ordered.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Indexierung und Slicing 

Sets können nicht indiziert werden, da sie nicht geordnet sind.

</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Manipulating sets 

Sets are mutable, so they have methods to add and remove elements.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Indexierung 

Sets sind veränderbar, d.h. sie haben Methoden, um Elemente hinzuzufügen und zu entfernen.

</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

### Adding elements    
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

### Elemente hinzufügen
</div>
</div>

#### `add()`

In [ ]:
# add a single element

set.add(3)
set

#### `update()`

In [ ]:
# add multiple elements from a collection

set.update([4,5,6])
set

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

### Removing elements    
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

### Elemente entfernen
</div>
</div>

#### `remove()`

In [ ]:
# remove a specific element

set.remove(4)
set

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

 `remove()` raises an error if the element you are attempting to remove is missing:
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

 `remove()` gibt einen Fehler aus, wenn das Element, das man zu entfernen versucht, fehlt:
</div>
</div>

In [ ]:
# remove a specific element

set.remove(7)
set

#### `discard()`

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

 `discard()` does not raise an error if the element you are attempting to remove is missing, therefore it allows you to safely remove elements without worrying about handling exceptions if the element isn't found.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

 `discard()` löst keine Fehlermeldung aus, wenn das Element, das Sie zu entfernen versuchen, fehlt. 
 Daher können Sie Elemente sicher entfernen, ohne sich um die Behandlung von Ausnahmen zu kümmern, wenn das Element nicht gefunden wird.
</div>
</div>

In [ ]:
# remove a specific element

set.discard(7)
set

#### `pop()`

In [ ]:
# remove and return an arbitrary element

print(set.pop())
print(set)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

Since sets are unordered, there is no "first" or "last" element, so `pop()` removes a random element.

Why is it useful? Unlike lists, which need to shift elements when removing items, sets don't require such reorganization, making `pop()` faster for large datasets.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Da Sets ungeordnet sind, gibt es kein "erstes" oder "letztes" Element, so dass `pop()` ein zufälliges Element entfernt.

Warum ist das nützlich? Im Gegensatz zu Listen, bei denen die Elemente beim Entfernen von Elementen verschoben werden müssen, ist bei Sets eine solche Reorganisation nicht erforderlich, wodurch `pop()` bei großen Datensätzen schneller ist.
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

**Sets** are useful in Python for several reasons:

  - **Uniqueness:** Sets automatically remove duplicate elements, ensuring that each item is unique. This makes them ideal for scenarios where duplicate values need to be filtered out.

  - **Efficient membership testing:** Sets provide fast checks for whether an element is present, using the `in` operator. This is much faster compared to lists, especially for large collections.

  - **Set operations:** Sets support mathematical set operations such as union, intersection, difference, and symmetric difference, making them useful for tasks involving comparisons between collections.

  - **Unordered collections:** Since sets are unordered, they can be more memory efficient and optimized for certain operations, such as searching for unique items in large datasets.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
**Sets** sind in Python aus mehreren Gründen nützlich:

  - **Einzigartigkeit:** Sets entfernen automatisch doppelte Elemente und stellen sicher, dass jedes Element eindeutig ist. Das macht sie ideal für Szenarien, in denen doppelte Werte herausgefiltert werden müssen.

  - **Effiziente Zugehörigkeitsprüfung:** Sets bieten schnelle Überprüfungen, ob ein Element vorhanden ist, indem sie den `in`-Operator verwenden. Dies ist im Vergleich zu Listen viel schneller, besonders bei großen Sammlungen.

  - **Mengenoperationen:** Mengen unterstützen mathematische Mengenoperationen wie Vereinigung, Schnittmenge, Differenz und symmetrische Differenz, was sie für Aufgaben mit Vergleichen zwischen Sammlungen nützlich macht.

  - **Ungeordnete Sammlungen:** Da Mengen ungeordnet sind, können sie speichereffizienter sein und für bestimmte Operationen optimiert werden, z. B. für die Suche nach eindeutigen Elementen in großen Datenbeständen.

</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Comparison between lists, tuples, and sets 
 Feeling overwhelmed? No need to worry! The table below summarizes the key differences between lists, tuples, and sets. There's no need to memorize them.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Vergleich zwischen Listen, Tupeln und Sets
Fühlen Sie sich überwältigt? Kein Grund zur Sorge! In der folgenden Tabelle sind die wichtigsten Unterschiede zwischen Listen, Tupeln und Sets zusammengefasst. Sie brauchen sie nicht auswendig zu lernen.
</div>
</div>

| Operation| Operation (German)   | Lists  | Tuples | Sets   |
|--------------------------|--------------------------|------------------------|------------------------|------------------------|
| Syntax   | Syntax   | `[ ]`  | `( )`  | `{ }`  |
| Mutability   | Veränderbarkeit  | &check;| &cross;| &check;|
| Ordered  | Geordnet | &check;| &check;| &cross;|
| Duplicates Allowed| Duplikate erlaubt| &check;| &check;| &cross;|
| Indexing | Indizierung  | &check;| &check;| &cross;|
| Slicing  | Slicing   | &check;| &check;| &cross;|
| Iteration| Iteration| &check;| &check;| &check;|
| Adding Elements  | Elemente hinzufügen  | `append()`, `insert()` | N/A| `add()`|
| Removing Elements | Elemente entfernen   | `remove()`, `pop()`| N/A| `remove()`, `discard()`, `pop()` |
| Concatenation| Verkettung   | `+`| `+`| N/A|
| Length   | Länge| `len()`| `len()`| `len()`|
| Membership Test  | Mitgliedschaftstest  | `in`   | `in`   | `in`   |
| Repetition   | Wiederholung | `*`| `*`| N/A|
| Conversion to List| Umwandlung zu Liste  | -  | `list()`   | `list()`   |
| Conversion to Tuple   | Umwandlung zu Tupel  | `tuple()`  | -  | `tuple()`  |
| Conversion to Set | Umwandlung zu Set| `set()`| `set()`| -  |

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Operators 
 Additionally, sets support the following operators: 

-  `-`: difference (NOT)

-  `ˆ`: symmetric difference (XOR)

-  `&`: intersection (AND)

-  `|`: union (OR)


</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Operatoren 
Außerdem unterstützen Sets die folgenden Operatoren:

-  `-`: Differenz

-  `ˆ`: symmetrische Differenz

-  `&`: Schnittmenge

-  `|`: Vereinigung


</div>
</div>

In [ ]:
set_1 = {'A', 'B', 'C', 'D'}

set_2 = {'C', 'D', 'X', 'Y', 'Z'}

In [ ]:
set_1 - set_2   # elements in set_1 but NOT in set_2

In [ ]:
set_2 - set_1   # elements in set_2 but NOT in set_1

In [ ]:
set_1 ^ set_2   # symmetric difference: elements that are not in both sets

In [ ]:
set_1 & set_2   # intersection: elements common to both sets

In [ ]:
set_1 | set_2   # union: all unique elements from both sets

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise
The crystal lattice structures of the first-row transition metal elements are given below. Some elements have different crystal structures under varying conditions of temperature and pressure. 

- fcc: Cu, Co, Fe, Mn, Ni, Sc

- bcc: Cr,Fe,Mn,Ti,V

- hcp: Co, Ni, Sc, Ti, Zn

Use Python sets to group them and determine the following:

  - Which metals only exist in face-centered cubic (**fcc**), body-centered cubic (**bcc**), or hexagonal close-packed (**hcp**) structures?

  - Which metals exist in only two of these structures?
  
  - Which metals do not form an **hcp** structure?
  
  - Do any of them exist in all three structures?

 Beware: The union operator has lower precedence than the difference operator, so you might parentheses.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung
Die Kristallgitterstrukturen der Elemente der ersten Reihe der Übergangsmetalle sind unten angegeben. Einige Elemente haben unter verschiedenen Temperatur- und Druckbedingungen unterschiedliche Kristallstrukturen. 

- fcc: Cu, Co, Fe, Mn, Ni, Sc

- bcc: Cr,Fe,Mn,Ti,V

- hcp: Co, Ni, Sc, Ti, Zn

Benutzen Sie Python-Sets, um sie zu gruppieren und folgendes zu bestimmen:

  - Welche Metalle kommen nur in kubisch-flächenzentrierten (**fcc**), kubisch-raumzentrierten (**bcc**) oder hexagonal dicht gepackten (**hcp**) Strukturen vor?
  
  - Welche Metalle kommen nur in zwei dieser Strukturen vor?
  
  - Welche Metalle bilden keine **hcp**-Struktur?
  
  - Kommt eines von ihnen in allen drei Strukturen vor?

Achtung: Der Vereinigungsoperator hat einen niedrigeren Vorrang als der Differenzoperator, daher können Sie Klammern setzen.
</div>
</div>

In [ ]:
#countdown
countdown_timer(20)

In [ ]:
fcc = {'Cu', 'Co', 'Fe', 'Mn', 'Ni', 'Sc'}
bcc = {'Cr', 'Fe', 'Mn', 'Ti', 'V'}
hcp = {'Co', 'Ni', 'Sc', 'Ti', 'Zn'}

#### 1

In [ ]:
# fcc only but NOT bcc and hcp
fcc - bcc - hcp

In [ ]:
# fcc only but NOT bcc and hcp
fcc - (bcc | hcp)

In [ ]:
# bcc only 
bcc - fcc - hcp

In [ ]:
# hcp only 
hcp - fcc - bcc

#### 2

In [ ]:
# fcc AND bcc but NOT hcp
fcc & bcc - hcp

In [ ]:
# fcc AND hcp but NOT bcc
fcc & hcp - bcc

In [ ]:
# bcc AND hcp but NOT fcc
bcc & hcp - fcc

#### 3

In [ ]:
# NOT hcp, but either fcc OR bcc
(fcc | bcc) - hcp   # the union operator has lower priority than the intersection

#### 4

In [ ]:
# in fcc AND bcc AND hcp
fcc & bcc & hcp

<div class="alert alert-block alert-warning">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

# Dictionaries 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

# Dictionaries
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

A Python dictionary is a type of associative array: A mapping of key objects to value objects. Instead of indexing with an integer, each value is associated with a unique key (which must be an immutable object).

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Ein Python-Dictionary ist eine Art assoziatives Array: Ein Mapping von Schlüsselobjekten auf Wertobjekte. Anstatt mit einer ganzen Zahl zu indizieren, ist jeder Wert mit einem eindeutigen Schlüssel verbunden (der ein unveränderliches Objekt sein muss).

</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Creating dictionaries 

A dictionary is created using curly braces `{}` with comma-separated `key:value` pairs. For example:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Dictionaries erstellen 

Ein Dictionary wird durch geschweifte Klammern `{}` mit kommagetrennten `key:value`-Paaren erstellt. Zum Beispiel:

</div>
</div>

In [ ]:
# The NATO phonetic alphabet is used to spell out words and names clearly over radio or telephone communication. 
# Here’s a Python dictionary that maps each letter of the alphabet to its corresponding phonetic code word:

nato_alphabet = {
'A': 'Alpha',
'B': 'Bravo',
'C': 'Charlie',
'D': 'Delta',
'E': 'Echo',
'F': 'Foxtrot',
'G': 'Golf',
'H': 'Hotel',
'I': 'India',
'J': 'Juliett',
'K': 'Kilo',
'L': 'Lima',
'M': 'Mike',
'N': 'November',
'O': 'Oscar',
'P': 'Papa',
'Q': 'Quebec',
'R': 'Romeo',
'S': 'Sierra',
'T': 'Tango',
'U': 'Uniform',
'V': 'Victor',
'W': 'Whiskey',
'X': 'X-ray',
'Y': 'Yankee',
'Z': 'Zulu'
}

# Print the dictionary
nato_alphabet

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Accessing and Modifying Data 

You can access values by their keys:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Zugriff auf und Änderung von Daten 

Sie können auf Werte über ihre Schlüssel zugreifen:

</div>
</div>

In [ ]:
nato_alphabet['Y']

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

To add or modify entries:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Um Einträge hinzuzufügen oder zu ändern:

</div>
</div>

In [ ]:
# modify one entry
nato_alphabet['Y'] = 'Yellow'

nato_alphabet

In [ ]:
# add a new entry
nato_alphabet['LL'] = 'Llamas'

nato_alphabet

In [ ]:
# remove an entry
del nato_alphabet['LL']

nato_alphabet

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise
 Create a Python dictionary that links the names of the first 10 elements of the periodic table to their symbols.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung
 Erstellen Sie ein Python-Wörterbuch, das die Namen der ersten 10 Elemente des Periodensystems mit ihren Symbolen verknüpft.

</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

In [ ]:
elements = {
"Hydrogen": "H",
"Helium": "He",
"Lithium": "Li",
"Beryllium": "Be",
"Boron": "B",
"Carbon": "C",
"Nitrogen": "N",
"Oxygen": "O",
"Fluorine": "F",
"Neon": "Ne"
}

print(elements)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

##  It gets interesting 

One application of a dictionary is as a simple database: Values can be stored in association with keys instead of assigned to individual variable names. 

This is much easier to maintain and has the advantage that the keys can be arbitrary strings (or other immutable data types) whereas variable names are constrained by the syntax of the Python language.

Now, let's assume we want to store more than just the element's symbol; we also want to include its atomic number and atomic weight for easy access.

A *nested* dictionary in Python is a dictionary that contains other dictionaries as its values, allowing for the organization of complex, hierarchical data structures within a single dictionary. 

Consider the following dictionary that stores some information about the first 8 elements of the periodic table:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Es wird interessant 

Eine Anwendung eines Wörterbuchs ist die einer einfachen Datenbank: Werte können in Verbindung mit Schlüsseln gespeichert werden, anstatt sie einzelnen Variablennamen zuzuordnen. 

Dies ist viel einfacher zu pflegen und hat den Vorteil, dass die Schlüssel beliebige Zeichenketten (oder andere unveränderliche Datentypen) sein können, während Variablennamen durch die Syntax der Sprache Python eingeschränkt sind.

Nehmen wir nun an, dass wir nicht nur das Symbol des Elements speichern wollen, sondern auch die Ordnungszahl und das Atomgewicht, um den Zugriff zu erleichtern. 

Ein *verschachteltes* Wörterbuch in Python ist ein Wörterbuch, das andere Wörterbücher als Werte enthält, was die Organisation komplexer, hierarchischer Datenstrukturen innerhalb eines einzigen Wörterbuchs ermöglicht. 

Betrachten Sie das folgende Wörterbuch, das einige Informationen über die ersten 8 Elemente des Periodensystems speichert:

</div>
</div>

In [3]:
elements = {
1: {"Name": "Hydrogen", "Symbol": "H", "Atomic Weight": 1.008},
2: {"Name": "Helium", "Symbol": "He", "Atomic Weight": 4.0026},
3: {"Name": "Lithium", "Symbol": "Li", "Atomic Weight": 6.94},
4: {"Name": "Beryllium", "Symbol": "Be", "Atomic Weight": 9.1022},
5: {"Name": "Boron", "Symbol": "B", "Atomic Weight": 10.81},
6: {"Name": "Carbon", "Symbol": "C", "Atomic Weight": 12.011},
7: {"Name": "Nitrogen", "Symbol": "N", "Atomic Weight": 14.007},
8: {"Name": "Oxygen", "Symbol": "O", "Atomic Weight": 15.999}
}

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

In this dictionary, the keys (`1`, `2`, `3`, etc.) represent the atomic numbers of the elements. 
Think of these as the row numbers in a table or spreadsheet. For example:

  -  Row 1 corresponds to Hydrogen.

  -  Row 2 corresponds to Helium.
  
  -  And so on.

Inside each element's dictionary, you have keys like `"Name"`, `"Symbol"`, and `"Atomic Weight"`. 
These are like the column headers in a table or spreadsheet:

  -  One column is labeled "Name".
  
  -   Another is labeled "Symbol".
  
  -   Another is labeled "Atomic Weight".

The values inside the inner dictionaries are the actual data in each cell of the table:

  -  For Hydrogen, the Name is "Hydrogen", the Symbol is "H", and the Atomic Weight is 1. 
  
  -  For Helium, the Name is "Helium", the Symbol is "He", and the Atomic Weight is 4.


</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

In diesem Wörterbuch stellen die Schlüssel (`1`, `2`, `3`, etc.) die Ordnungszahlen der Elemente dar.
Stellen Sie sich diese wie die Zeilennummern in einer Tabelle oder einem Arbeitsblatt vor.  Zum Beispiel:

  -  Zeile 1 entspricht dem Wasserstoff.
  
  -  Zeile 2 steht für Helium.
  
  -  Und so weiter.

Innerhalb des Wörterbuchs eines jeden Elements gibt es Schlüssel wie "Name", "Symbol" und "Atomgewicht".
Diese sind wie die Spaltenüberschriften in einer Tabelle oder Tabellenkalkulation:

  -  Eine Spalte ist mit "Name" beschriftet.
  
  -  Eine andere ist mit "Symbol" beschriftet.
  
  -  Eine weitere ist mit "Atomgewicht" beschriftet.

Die Werte in den inneren Wörterbüchern sind die tatsächlichen Daten in jeder Zelle der Tabelle:

  -  Für Wasserstoff ist der Name "Hydrogen", das Symbol ist "H" und das Atomgewicht ist 1.
  
  -  Für Helium ist der Name "Helium", das Symbol ist "He" und das Atomgewicht ist 4.


</div>
</div>

<div style="text-align: center;">
<img src="./figures/dict_table.png" width="350">
</div>

In [ ]:
# Example: Accessing information about Oxygen (Atomic Number 8)
elements[8]
#print(f"Element: {oxygen_info['Name']}, Symbol: {oxygen_info['Symbol']}, Atomic Weight: {oxygen_info['Atomic Weight']}")

In [ ]:
elements[8]["Atomic Weight"]

In [ ]:
molar_mass_glucose = (6 * elements[6]["Atomic Weight"]) + (12 * elements[1]["Atomic Weight"]) + (6 * elements[8]["Atomic Weight"])
molar_mass_glucose

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

Note that, in practice, this sort of data might be better held in a pandas DataFrame (my personal favorite, and the second part of this course.)

- Use dictionaries when you're working with small datasets, need fast lookups, or are storing simple, structured data (we are going to see a notable example in a second).

- Use Pandas DataFrames when you're working with large datasets, need to perform complex data analysis or manipulation, or require multi-dimensional data handling.


</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Beachten Sie, dass diese Art von Daten in der Praxis besser in einem Pandas DataFrame gespeichert werden sollte (mein persönlicher Favorit und der zweite Teil dieses Kurses).

- Verwenden Sie Dictionaries, wenn Sie mit kleinen Datensätzen arbeiten, schnelle Nachschlagewerke benötigen oder einfache, strukturierte Daten speichern (wir werden in einer Sekunde ein bemerkenswertes Beispiel sehen).

- Verwenden Sie Pandas DataFrames, wenn Sie mit großen Datensätzen arbeiten, komplexe Datenanalysen oder -manipulationen durchführen müssen oder eine mehrdimensionale Datenverarbeitung benötigen.


</div>
</div>

In [4]:
import pandas as pd 

df = pd.DataFrame(elements).T

df

Name Symbol Atomic Weight
1   Hydrogen      H         1.008
2     Helium     He        4.0026
3    Lithium     Li          6.94
4  Beryllium     Be        9.1022
5      Boron      B         10.81
6     Carbon      C        12.011
7   Nitrogen      N        14.007
8     Oxygen      O        15.999

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.5;">

###  Example: Physical Constants (SciPy)

The package `scipy.constants` provides the numerical values of many of the constants used in physics and chemistry.

Importing `scipy.constants` provides a dictionary, `scipy.constants.physical_constants` of *tuples* in the format: `(value, unit, uncertainty)`, keyed by identifying strings.

</div>
<div style="width: 48%; line-height: 1.5;color: grey;">

###  Beispiel: Physikalische Konstanten (SciPy)

Das Paket `scipy.constants` liefert die numerischen Werte vieler Konstanten, die in Physik und Chemie verwendet werden.

Der Import von `scipy.constants` liefert ein Wörterbuch, `scipy.constants.physical_constants` von *Tupeln* im Format: `(Wert, Einheit, Unsicherheit)`, verschlüsselt durch identifizierende Zeichenketten.

</div>
</div>

In [4]:
import scipy.constants as pc# we give it a short alias for convenience

In [6]:
pc.physical_constants  # behold the dictionary

{'Wien displacement law constant': (0.0028977685, 'm K', 5.1e-09),
 'atomic unit of 1st hyperpolarizablity': (3.20636151e-53,
  'C^3 m^3 J^-2',
  2.8e-60),
 'atomic unit of 2nd hyperpolarizablity': (6.2353808e-65,
  'C^4 m^4 J^-3',
  1.1e-71),
 'atomic unit of electric dipole moment': (8.47835309e-30, 'C m', 7.3e-37),
 'atomic unit of electric polarizablity': (1.648777274e-41,
  'C^2 m^2 J^-1',
  1.6e-49),
 'atomic unit of electric quadrupole moment': (4.48655124e-40,
  'C m^2',
  3.9e-47),
 'atomic unit of magn. dipole moment': (1.8548019e-23, 'J T^-1', 1.6e-30),
 'atomic unit of magn. flux density': (235051.756758, 'T', 7.1e-05),
 'deuteron magn. moment': (4.33073482e-27, 'J T^-1', 3.8e-34),
 'deuteron magn. moment to Bohr magneton ratio': (0.0004669754567, '', 5e-12),
 'deuteron magn. moment to nuclear magneton ratio': (0.8574382329,
  '',
  9.2e-09),
 'deuteron-electron magn. moment ratio': (-0.0004664345548, '', 5e-12),
 'deuteron-proton magn. moment ratio': (0.3070122084, '', 4.5

In [8]:
pc.physical_constants['Wien displacement law constant']  # access a constant by its identifying string

(0.0028977685, 'm K', 5.1e-09)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

Too much information? The most common constants, in SI units, can be imported directly:

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Zu viele Informationen? Die gängigsten Konstanten in SI-Einheiten können direkt importiert werden:

</div>
</div>

In [10]:
from scipy.constants import e, c, N_A, Boltzmann

print(e)
print(c)
print(N_A)
print(Boltzmann)

1.602176634e-19
299792458.0
6.02214076e+23
1.380649e-23


<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

*From now on, there's no need to manually assign variables to constants—you can simply import them directly!*

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

*Von nun an müssen Sie den Konstanten keine Variablen mehr manuell zuweisen, sondern können sie einfach direkt importieren!*

</div>
</div>

| Constant Name| Description (English)  | Description (German)  | Value | Units|
|--------------------------|----------------------------------------|-------------------------------------------|------------------------------------|------------------|
| `c`  | Speed of light in vacuum   | Lichtgeschwindigkeit im Vakuum| 299,792,458| m/s  |
| `G`  | Gravitational constant | Gravitationskonstante | 6.67430 × 10⁻¹¹| m³/kg/s² |
| `h`  | Planck's constant  | Plancksches Wirkungsquantum   | 6.62607015 × 10⁻³⁴ | J·s  |
| `k` or `Boltzmann`   | Boltzmann constant | Boltzmann-Konstante   | 1.380649 × 10⁻²³   | J/K  |
| `e`  | Elementary charge  | Elementarladung   | 1.602176634 × 10⁻¹⁹| C|
| `epsilon_0`  | Permittivity of free space | Elektrische Feldkonstante (Vakuum)| 8.8541878128 × 10⁻¹²   | F/m  |
| `mu_0`   | Permeability of free space | Magnetische Feldkonstante (Vakuum)| 1.25663706212 × 10⁻⁶   | N/A² |
| `N_A` or `Avogadro`  | Avogadro constant  | Avogadro-Konstante| 6.02214076 × 10²³  | mol⁻¹|
| `R`  | Gas constant   | Gaskonstante  | 8.314462618| J/mol/K  |
| `g`  | Standard acceleration due to gravity   | Standardfallbeschleunigung| 9.80665| m/s² |
| `m_e`| Electron mass  | Elektronenmasse   | 9.10938356 × 10⁻³¹ | kg   |
| `m_p`| Proton mass| Protonenmasse | 1.67262192369 × 10⁻²⁷  | kg   |
| `sigma` or `Stefan_Boltzmann` | Stefan-Boltzmann constant| Stefan-Boltzmann-Konstante| 5.670374419 × 10⁻⁸ | W/m²/K⁴  |
| `alpha`  | Fine-structure constant| Feinstrukturkonstante | 7.2973525693 × 10⁻³| /|
| `Rydberg`| Rydberg constant   | Rydberg-Konstante | 10,973,731.56816   | m⁻¹  |


<div class="alert alert-block alert-info">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

# Summary 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

# Zusammenfassung
</div>
</div>

| Operation | Beschreibung | Lists  | Arrays (NumPy)| Dictionaries   |
|-------------------------------|----------------------------------|------------------------------------|--------------------------------------|--------------------------------|
| **Creation**  | **Erstellung**   | `my_list = [1, 2, 3]`  | `import numpy as np``my_array = np.array([1, 2, 3])` | `my_dict = {'a': 1, 'b': 2}`  |
| **Access Element**| **Elementzugriff**   | `my_list[0]`| `my_array[0]`| `my_dict['a']`  |
| **Add Element**   | **Element hinzufügen**   | `my_list.append(4)`| `np.append(my_array, 4)`  | `my_dict['c'] = 3`  |
| **Insert Element**| **Element einfügen** | `my_list.insert(1, 10)`| `np.insert(my_array, 1, 10)`  | N/A  |
| **Remove Element**| **Element entfernen**| `my_list.remove(2)`| `np.delete(my_array, 1)`  | `del my_dict['a']`  |
| **Update Element**| **Element aktualisieren**| `my_list[0] = 10`  | `my_array[0] = 10`| `my_dict['a'] = 10` |
| **Length**| **Länge**| `len(my_list)` | `len(my_array)`   | `len(my_dict)`  |
| **Slice** | **Slicing**  | `my_list[1:3]` | `my_array[1:3]`   | N/A  |
| **Mathematical Operations**   | **Mathematische Operationen**| Not directly supported | `my_array * 2`| N/A  |
| **Sort**  | **Sortieren**| `my_list.sort()`   | `np.sort(my_array)`  | N/A  |
